### PR TITLE
#57 alter table support for constraints

### PR DESCRIPTION
**Pull request by kriti-sc**
This PR aims to create names for constraints.

### CODE REVIEW

**Comment by kriti-sc** [comment]
One of the requirements would be to allow renaming constraints. Do we need a separate function for that or the process of renaming would follow the same procedure as constraint creation?

File: src/catalog/catalog_entry/duck_constraint_entry.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+DuckConstraintEntry::DuckConstraintEntry(Catalog &catalog, SchemaCatalogEntry &schema, CreateConstraintInfo &info)
+    : ConstraintCatalogEntry(catalog, schema, info) {
+}
+
+DuckConstraintEntry::~DuckConstraintEntry() {
+}
+
+string DuckConstraintEntry::GetSchemaName() const {
+	return "";
```
Comment (kriti-sc) [inline src/catalog/catalog_entry/duck_constraint_entry.cpp:14]: I am adding constraints as currently indexes are added. 

Indexes have this `shared_ptr<DataTableInfo> info` variable, which is contains schema name and table name like so: `info->schema`, `info->table`. 

How should I do the same for constraints?

**Review by lnkuiper — CHANGES_REQUESTED** [review]
The general flow of creating an index is:
1. Parser: `CREATE INDEX my_index ON tbl(i);` (Creates a Postgres parse tree)
  * CREATE statements are found in `third_party/libpg_query/grammar/statements/create.y`
  * When modifying the grammar run `python scripts/generate_grammar.py`, when modifying the lexer, run `python scripts/generate_flex.py`
2. Transformer (Transforms from Postgres parse tree to DuckDB parse tree)
  * For `CREATE INDEX`, you can find this here: `src/parser/transform/statement/transform_create_index.cpp`
3. Binder (Binds the table/column names to actual tables/columns, creating a logical plan)
  * For `CREATE INDEX`, you can find this in `src/planner/binder/statement/bind_create.cpp at line 462`
4. Planner (Create a physical plan from the logical plan)
  * For `CREATE INDEX`, this is done here: `src/execution/physical_plan/plan_create_index.cpp`
5. Execution (Executes a physical plan)
  * Done here: `src/execution/operator/schema/physical_create_index.cpp`

Creating/dropping a constraint should follow the same flow, and you can roughly copy the steps since the index is also a catalog object.
The physical creation of an index is much more involved though, since it requires scanning the table.

Maybe you can to do this flow for a simple constraint such as `NOT NULL`:
`ALTER TABLE tbl ADD CONSTRAINT constraint_name NOT NULL (column1)`

Checking for non-null does not require scanning the table, as we have statistics that can tell us whether a column contains `NULL`.
If the column contains `NULL`, this should fail. If not, the constraint should be added, and subsequent appends of `NULL` should fail.

For example, you can write a test like this:
```sql
statement ok
CREATE TABLE test (i INT);

statement ok
ALTER TABLE test ADD CONSTRAINT constraint_name NOT NULL (i);

statement error
INSERT INTO test values (NULL);
```
You should also try these with `START TRANSACTION` / `COMMIT`. You can run tests by doing `make debug` and then `./build/debug/test/unittest test/sql/path/to/my/test.test`.
Tests are run automatically on your own fork: https://github.com/kriti-sc/duckdb/actions


Similarly, constraints like `CHECK (i > 10)` can also be checked through statistics.
You can simply throw a `NotImplementedException` for contraints that we don't support in this way yet.

I don't think our statistics are transactional (if a `NULL` was added and then deleted, then the stats will still say that the column contains `NULL`), but that's OK for now, we can solve that later.

Edit: `ALTER` should go through `transform_alter_table.cpp`, etc. this is already in place, just needs to have support for added for renaming constraints.

**Comment by hannes** [comment]
@kriti-sc can you address @lnkuiper's comments or close please?

**Comment by kriti-sc** [comment]
I am working on the comments @hannes.

**Comment by erikhuck** [comment]
@kriti-sc @lnkuiper  Any updates here?

**Comment by Mytherin** [comment]
Closing due to inactivity

**Pull request closed without merging**